jobs/dtd-metrics/src/services/tokens-store.service.ts

```diff
@@ -0,0 +1,55 @@
+import { z } from 'zod'
+import { AthenaClientService } from './athena-client.service.js'
+import { Row } from '@aws-sdk/client-athena'
+import { env } from '../configs/env.js'
+import { aggregateTokensCount } from '../utils/helpers.utils.js'
+
+export const TokensByDay = z.array(z.object({ day: z.date(), tokens: z.number() }))
+export type TokensByDay = z.infer<typeof TokensByDay>
+
+export class TokensStore {
+  private static athena = new AthenaClientService({ outputLocation: `s3://${env.ATHENA_OUTPUT_BUCKET}` })
+
+  private static instance: TokensStore
+  public tokensByDay: TokensByDay
+  public totalTokens: number
+
+  private constructor(tokensByDay: TokensByDay) {
+    this.tokensByDay = tokensByDay
+    this.totalTokens = aggregateTokensCount(tokensByDay)
+  }
+
+  public static async getInstance(): Promise<TokensStore> {
+    if (!TokensStore.instance) {
+      const tokensByDay = await this.getTokensGroupedByDay()
```

Comment:
I agree 100% with you on this. But I thought of the singleton being a good fit here not just for the easy cache, but also for DX reasons.

The retrieval of the tokens is an expensive operation and the singleton starts that operation only when it is instantiated. Meaning that if I want to produce only a specific metric that does not use the tokens data, I don't have to wait for the token retrieving operation since the singleton will not be instantiated, speeding up dev time.

The retrieval of the tokens is a lot faster with the introduction of Athena, so this might be an outdated solution.
Thanks for the heads up, I'll consider it for the future.

Reply:
it makes totally sense, we can further discuss this topic but no blockers for this PR 👍